### PR TITLE
MAINT: optimize: let user choose convergence criterion in `newton`

### PR DESCRIPTION
**Pull request by person142**
**Edit**:  before `newton` could only use absolute step size for its convergence
criterion; this lets it also use relative step size, absolute size of
the function value, or relative size of the function value.

~~Before `newton` used step size for its convergence criterion; now it
uses the absolute value of the objective function.~~ Closes gh-6379.

### CODE REVIEW

**Comment by pv** [comment]
For the newton method, the step size may be a reasonable termination condition, so I think this is a problem only for the secant method (when fprime is not given).

**Comment by person142** [comment]
Yes... both conditions get used. It's also possible to introduce an extra keyword argument `ftol` and allow either step size or objective function size to be used as a stopping criterion; `tol=None` meaning "don't use step size". That also has the advantage of being back-compatible.

**Comment by person142** [comment]
i.e. something like this. If @pv is anti I'll change how `tol` works for just the secant method. Will need to do some squashing/cherry picking either way.

**Comment by dlax** [comment]
Commit https://github.com/scipy/scipy/pull/6380/commits/617b3d675cbfe28e8f4311767b13d6c7f5d56ced is unrelated and should be submitted independently I think.

File: scipy/optimize/zeros.py [hunk 1]
```diff
@@ -112,36 +129,47 @@ def newton(func, x0, fprime=None, args=(), tol=1.48e-8, maxiter=50,
...
+    if tol is None and ftol is None:
+        raise ValueError("tol and ftol cannot both be None")
+    elif tol is not None and ftol is not None:
+        raise ValueError("only one of tol and ftol can not be None")
+    elif tol is not None and tol <= 0:
         raise ValueError("tol too small (%g <= 0)" % tol)
-    if maxiter < 1:
+    elif ftol is not None and ftol <= 0:
+        raise ValueError("tol too small (%g <= 0)" % ftol)
+    elif maxiter < 1:
```
Comment (dlax) [inline scipy/optimize/zeros.py]: check on `maxiter < 1` should remain an `if`

**Comment by person142** [comment]
Made some changes:

- Added a couple more possible convergence criteria which are selected with a `halt` keyword. (Removed the `ftol` stuff; it was ugly.)
- Removed the changes in `signal` (which were no longer necessary because this version is backwards compatible).
- Rebased and squashed.

File: scipy/optimize/zeros.py [hunk 2]
```diff
@@ -62,11 +62,22 @@ def results_c(full_output, r):
...
 
+def _newton_halt(p, dp, fp, tol, halt):
+    if halt == 'astep':
+        res = True if abs(dp) <= tol else False
+    elif halt == 'rstep':
+        res = True if abs(dp) <= tol*abs(p) else False
+    elif halt == 'afval':
+        res = True if abs(fp) <= tol else False
+    else:
+        res = True if abs(fp) <= tol*abs(p) else False
```
Comment (dlax) [inline scipy/optimize/zeros.py]: all these `res = True if <something> else False` could be written `res = <statement>` since `<statement>` already evaluates as a boolean value.

also this could also just be a dict like:
```
return {
    'astep': abs(dp) <= tol,
    'rstep': abs(dp) <= tol*abs(p),
    ....
}[halt]
````

File: scipy/optimize/zeros.py [hunk 3]
```diff
@@ -112,36 +134,48 @@ def newton(func, x0, fprime=None, args=(), tol=1.48e-8, maxiter=50,
...
+
     if fprime is not None:
-        # Newton-Rapheson method
-        # Multiply by 1.0 to convert to floating point.  We don't use float(x0)
+        # Newton-Raphson method
+        # Multiply by 1.0 to convert to floating point. We don't use float(x0)
         # so it still works if x0 is complex.
         p0 = 1.0 * x0
+        myargs = (p0,) + args
+        fval = func(*myargs)
```
Comment (dlax) [inline scipy/optimize/zeros.py]: It's unclear to me why you moved these lines ouside the loop and then repeat them at the end. Could you explain?

File: scipy/optimize/tests/test_zeros.py [hunk 4]
```diff
@@ -50,18 +50,23 @@ def test_newton(self):
...
+                x = zeros.newton(f, 3, halt=halt)
+                assert_allclose(f(x), 0, atol=1e-6)
+                x = zeros.newton(f, 3, fprime=f_1, halt=halt)
+                assert_allclose(f(x), 0, atol=1e-6)
+                x = zeros.newton(f, 3, fprime=f_1, fprime2=f_2, halt=halt)
+                assert_allclose(f(x), 0, atol=1e-6)
 
     def test_deriv_zero_warning(self):
-        func = lambda x: x**2
+        func = lambda x: x**2 + 1
```
Comment (dlax) [inline scipy/optimize/tests/test_zeros.py]: unrelated change?

**Review by dlax — CHANGES_REQUESTED** [review]
Overall this looks good, modulo minor comments inline.

File: scipy/optimize/zeros.py (hunk 3 above)
Comment (person142) [inline scipy/optimize/zeros.py]: I might be thinking through this incorrectly, but the idea was that the previous sequence of events in the loop was:

compute function -> take a step -> check halting condition

which worked fine because the halting condition didn't involve the current function value. With these changes it could, so the sequence in the loop was modified to be:

take step -> compute function at new point -> check halting condition.

Comment (dlax) [inline scipy/optimize/zeros.py]: Ok. Wouldn't `compute function -> check halting condition -> take a step` work then?

File: scipy/optimize/zeros.py [hunk 5]
```diff
@@ -112,36 +134,48 @@ def newton(func, x0, fprime=None, args=(), tol=1.48e-8, maxiter=50,
...
         fder2 = 0
         for iter in range(maxiter):
-            myargs = (p0,) + args
             fder = fprime(*myargs)
             if fder == 0:
```
Comment (dlax) [inline scipy/optimize/zeros.py]: Also, maybe it'd be clearer if this clause was included in `_newton_halt` check, so as to keep only one single exit point in the loop/function.

File: scipy/optimize/zeros.py (hunk 3 above)
Comment (person142) [inline scipy/optimize/zeros.py]: I don't think it would work on the first iteration--there wouldn't be a step size yet.

Comment (dlax) [inline scipy/optimize/zeros.py]: Indeed. But if the first function evaluation satisfies a termination condition (one of `*fval` for instance), there's no reason to perform another function evaluation. Also, upon first iteration two function evaluations are performed with only one check of termination condition so this is different from before (one more function evaluation) for no apparent reason as far as the default `astep` termination condition is concerned. So maybe this means that there cannot be a single point to check the termination condition for, but two, depending on whether `halt` is `*step` or `*fval`.

**Comment by rgommers** [comment]
@person142 looks like this is close to done, maybe good to finish it up?

**Comment by perimosocordiae** [comment]
@person142 did you mean to close this?

**Comment by person142** [comment]
No... perhaps I deleted the branch by accident and that auto-closes the PR?
